Kytos shutdown right after successful return to user request can leads to data loss

Due to the async way the `topology/storehouse.py` persists data, the user request can result in a successful return while data is not guaranteed to be persisted. If Kytos receives a shutdown request right after sending the successful return to the user, user data may be lost.

As an example, let's take the following end-to-end tests that are running in a daily basis and eventually fails (very rare):
```
=================================== FAILURES ===================================
_____________ TestE2ETopology.test_030_disabling_switch_persistent _____________

self = <tests.test_e2e_05_topology.TestE2ETopology instance at 0x7f4b114c8050>

   def test_030_disabling_switch_persistent(self):
       """
       Test /api/kytos/topology/v3/switches/{dpid}/disable on POST
       supported by
           /api/kytos/topology/v3/switches on GET
       """

       switch_id = "00:00:00:00:00:00:00:01"

       # Enable the switch
       api_url = KYTOS_API + '/topology/v3/switches/%s/enable' % switch_id
       requests.post(api_url)

       # Disable the switch
       api_url = KYTOS_API + '/topology/v3/switches/%s/disable' % switch_id
       response = requests.post(api_url)
       assert response.status_code == 201

       self.restart()

       # Check if the switch is disabled
       api_url = KYTOS_API + '/topology/v3/switches'
       response = requests.get(api_url)
       data = response.json()
     assert data['switches'][switch_id]['enabled'] is False
E       assert True is False

tests/test_e2e_05_topology.py:154: AssertionError
```

Running the same test manually revels that the fail occurs in a very rare rate:
```
# cat run-test-topo
( for i in $(seq 1 100); do echo $i $(date); python -m pytest tests/test_e2e_05_topology.py::TestE2ETopology::test_030_disabling_switch_persistent; done; date ) | tee log-topo-e2e-$(date +%Y%m%d%H%M%S)
# sh run-test-topo
# sh run-test-topo
# sh run-test-topo
# sh run-test-topo
# for cada in log-topo-e2e-20211218075644 log-topo-e2e-20211218111711 log-topo-e2e-20211218130921 log-topo-e2e-20211218150530; do echo $cada; egrep -o " (passed|failed) " $cada | sort | uniq -c; done
log-topo-e2e-20211218075644
      2  failed
     98  passed
log-topo-e2e-20211218111711
      2  failed
     98  passed
log-topo-e2e-20211218130921
      3  failed
     97  passed
log-topo-e2e-20211218150530
    100  passed
```

In summary, 7 failures out of 400 repetitions of the same test.

Analyzing more deeply those requests (kytos configured in DEBUG mode) give us more clue about what is happening. Before showing the failed cases, let's take a look on one of the successful cases (passed):

```
## To generate the log file containing only the subset of log messages for the tests, we use the date/time when the test began and finished:
# tail -n 11 log-topo-e2e-20211218150530
100 Sat Dec 18 16:56:23 UTC 2021
============================= test session starts ==============================
platform linux2 -- Python 2.7.16, pytest-3.10.1, py-1.7.0, pluggy-0.8.0
rootdir: /kytos-end-to-end-tests, inifile:
plugins: timeout-1.3.3
collected 1 item

tests/test_e2e_05_topology.py .                                          [100%]

========================== 1 passed in 66.05 seconds ===========================
Sat Dec 18 16:57:29 UTC 2021

# egrep "Dec 18 16:56:[2345]|Dec 18 16:57" /var/log/syslog > /tmp/log-passed

## filter out by the subset of messages that are important for this issue (line numbers inserted to facilitate comprehension):
# egrep "(Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled|POST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable|Added: kytos.storehouse.update|Removed: kytos.storehouse.update|looking for listeners for kytos.storehouse.update|Network administrative status saved in kytos.topology.status|kytos.core.controller:DEBUG controller:283: .* threads before loop.stop)" /tmp/log-passed  | sed -n '/Storing administrative state from switch/,/threads before loop.stop/ p' | awk '{print NR"\t"$0}'
1	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,891 (Thread-13) kytos.napps.kytos/topology:INFO main:273:  Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled.
2	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,900 (MainThread) kytos.core.buffers:DEBUG buffers:101:  [buffer: app_event] Removed: kytos.storehouse.update
3	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,900 (MainThread) kytos.core.controller:DEBUG controller:495:  looking for listeners for kytos.storehouse.update
4	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,906 (Thread-13) kytos.core.buffers:DEBUG buffers:46:  [buffer: app_event] Added: kytos.storehouse.update
5	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,908 (Thread-13) werkzeug:INFO _internal:113:  127.0.0.1 - - [18/Dec/2021 16:57:07] "#033[37mPOST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable HTTP/1.1#033[0m" 201 -
6	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,919 (ThreadPoolExecutor-0_15) kytos.napps.kytos/topology:INFO storehouse:121:  Network administrative status saved in kytos.topology.status.9380b69bd76640c59804cdc0d9da5bf1
7	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,926 (MainThread) kytos.core.buffers:DEBUG buffers:101:  [buffer: app_event] Removed: kytos.storehouse.update
8	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,927 (MainThread) kytos.core.controller:DEBUG controller:495:  looking for listeners for kytos.storehouse.update
9	Dec 18 16:57:07 a23139f5c9dc 2021-12-18 16:57:07,932 (ThreadPoolExecutor-0_7) kytos.napps.kytos/topology:INFO storehouse:121:  Network administrative status saved in kytos.topology.status.9380b69bd76640c59804cdc0d9da5bf1
10	Dec 18 16:57:08 a23139f5c9dc 2021-12-18 16:57:08,542 (MainThread) kytos.core.controller:DEBUG controller:283:  9 threads before loop.stop: [<_MainThread(MainThread, started 140616468961088)>, <InotifyObserver(Thread-1, started daemon 140616410949376)>, <Thread(APScheduler, started daemon 140616401770240)>, <Thread(APScheduler, started daemon 140616387712768)>, <Thread(ThreadPoolExecutor-4_0, started daemon 140615987738368)>, <Thread(ThreadPoolExecutor-4_1, started daemon 140615970952960)>, <Thread(ThreadPoolExecutor-4_2, started daemon 140615819982592)>, <Thread(ThreadPoolExecutor-4_3, started daemon 140615803197184)>, <Thread(ThreadPoolExecutor-4_4, started daemon 140615794804480)>]
```

Line 1 is called right before persisting user request and returning to the user (https://github.com/kytos-ng/topology/blob/master/main.py#L273). Line 2 and 3 we can ignore because they belong to the previous request. Line 4 is called when the storehouse event is created to persist data. Line 5 is the return to the user request. Line 6 we can ignore because it also belongs to previous request. Line 7 and 8 are Kytos Controller handling the queued event to persist data. Line 9 is from the `topology/storehouse.py` confirming the persisted data (https://github.com/kytos-ng/topology/blob/master/storehouse.py#L114). Line 10 is the kytos shutdown process (hard shutdown, no further action/thread is executed).

Now, let's take a look on the same analysis when a failure occurs. I've analyzed 3 cases from log `log-topo-e2e-20211218130921`, but all other files present similar behavior:

```
### Fail 1:
# egrep "^Dec 18 14:17:3[6789]|^Dec 18 14:17:[45]|^Dec 18 14:18:[0123]|^Dec 18 14:18:4[01234]" /var/log/syslog > /tmp/log-failed-60
# egrep "(Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled|POST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable|Added: kytos.storehouse.update|Removed: kytos.storehouse.update|looking for listeners for kytos.storehouse.update|Network administrative status saved in kytos.topology.status|kytos.core.controller:DEBUG controller:283: .* threads before loop.stop)" /tmp/log-failed-60  | sed -n '/Storing administrative state from switch/,/threads before loop.stop/ p' | awk '{print NR"\t"$0}'
1	Dec 18 14:18:23 a23139f5c9dc 2021-12-18 14:18:23,637 (Thread-13) kytos.napps.kytos/topology:INFO main:273:  Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled.
2	Dec 18 14:18:23 a23139f5c9dc 2021-12-18 14:18:23,649 (Thread-13) kytos.core.buffers:DEBUG buffers:46:  [buffer: app_event] Added: kytos.storehouse.update
3	Dec 18 14:18:23 a23139f5c9dc 2021-12-18 14:18:23,651 (Thread-13) werkzeug:INFO _internal:113:  127.0.0.1 - - [18/Dec/2021 14:18:23] "#033[37mPOST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable HTTP/1.1#033[0m" 201 -
4	Dec 18 14:18:24 a23139f5c9dc 2021-12-18 14:18:24,209 (MainThread) kytos.core.controller:DEBUG controller:283:  9 threads before loop.stop: [<_MainThread(MainThread, started 140532382791488)>, <InotifyObserver(Thread-1, started daemon 140532324779776)>, <Thread(APScheduler, started daemon 140532315600640)>, <Thread(APScheduler, started daemon 140532301543168)>, <Thread(ThreadPoolExecutor-4_0, started daemon 140531772933888)>, <Thread(ThreadPoolExecutor-4_1, started daemon 140531756148480)>, <Thread(ThreadPoolExecutor-4_2, started daemon 140531747755776)>, <Thread(ThreadPoolExecutor-4_3, started daemon 140531657074432)>, <Thread(ThreadPoolExecutor-4_4, started daemon 140531648681728)>]

### Fail 2:
# egrep "^Dec 18 14:22:1[3-9]|^Dec 18 14:22:[2-5]|^Dec 18 14:23:[01]|^Dec 18 14:23:2[012]" /var/log/syslog > /tmp/log-failed-64
# egrep "(Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled|POST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable|Added: kytos.storehouse.update|Removed: kytos.storehouse.update|looking for listeners for kytos.storehouse.update|Network administrative status saved in kytos.topology.status|kytos.core.controller:DEBUG controller:283: .* threads before loop.stop)" /tmp/log-failed-64  | sed -n '/Storing administrative state from switch/,/threads before loop.stop/ p' | awk '{print NR"\t"$0}'
1	Dec 18 14:22:59 a23139f5c9dc 2021-12-18 14:22:59,605 (Thread-13) kytos.napps.kytos/topology:INFO main:273:  Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled.
2	Dec 18 14:22:59 a23139f5c9dc 2021-12-18 14:22:59,624 (Thread-13) kytos.core.buffers:DEBUG buffers:46:  [buffer: app_event] Added: kytos.storehouse.update
3	Dec 18 14:22:59 a23139f5c9dc 2021-12-18 14:22:59,626 (Thread-13) werkzeug:INFO _internal:113:  127.0.0.1 - - [18/Dec/2021 14:22:59] "#033[37mPOST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable HTTP/1.1#033[0m" 201 -
4	Dec 18 14:23:00 a23139f5c9dc 2021-12-18 14:23:00,293 (MainThread) kytos.core.controller:DEBUG controller:283:  8 threads before loop.stop: [<_MainThread(MainThread, started 140693144561472)>, <Thread(APScheduler, started daemon 140693077370624)>, <Thread(APScheduler, started daemon 140692860958464)>, <Thread(ThreadPoolExecutor-4_0, started daemon 140692803000064)>, <Thread(ThreadPoolExecutor-4_1, started daemon 140692517021440)>, <Thread(ThreadPoolExecutor-4_2, started daemon 140692508628736)>, <Thread(ThreadPoolExecutor-4_3, started daemon 140692491843328)>, <Thread(ThreadPoolExecutor-4_4, started daemon 140692483450624)>]


### Fail 3:
# egrep "^Dec 18 14:43:2[89]|^Dec 18 14:43:[3-5]|^Dec 18 14:44:[0123]|^Dec 18 14:44:4[0]" /var/log/syslog > /tmp/log-failed-82
# egrep "(Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled|POST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable|Added: kytos.storehouse.update|Removed: kytos.storehouse.update|looking for listeners for kytos.storehouse.update|Network administrative status saved in kytos.topology.status|kytos.core.controller:DEBUG controller:283: .* threads before loop.stop)" /tmp/log-failed-82  | sed -n '/Storing administrative state from switch/,/threads before loop.stop/ p' | awk '{print NR"\t"$0}'
1	Dec 18 14:44:19 a23139f5c9dc 2021-12-18 14:44:19,012 (Thread-14) kytos.napps.kytos/topology:INFO main:273:  Storing administrative state from switch 00:00:00:00:00:00:00:01 to disabled.
2	Dec 18 14:44:19 a23139f5c9dc 2021-12-18 14:44:19,018 (Thread-13) kytos.core.buffers:DEBUG buffers:46:  [buffer: app_event] Added: kytos.storehouse.update
3	Dec 18 14:44:19 a23139f5c9dc 2021-12-18 14:44:19,027 (Thread-14) kytos.core.buffers:DEBUG buffers:46:  [buffer: app_event] Added: kytos.storehouse.update
4	Dec 18 14:44:19 a23139f5c9dc 2021-12-18 14:44:19,028 (Thread-14) werkzeug:INFO _internal:113:  127.0.0.1 - - [18/Dec/2021 14:44:19] "#033[37mPOST /api/kytos/topology/v3/switches/00:00:00:00:00:00:00:01/disable HTTP/1.1#033[0m" 201 -
5	Dec 18 14:44:19 a23139f5c9dc 2021-12-18 14:44:19,617 (MainThread) kytos.core.controller:DEBUG controller:283:  8 threads before loop.stop: [<_MainThread(MainThread, started 139845325948736)>, <Thread(APScheduler, started daemon 139845191587584)>, <Thread(APScheduler, started daemon 139845177554688)>, <Thread(ThreadPoolExecutor-4_0, started daemon 139844781065984)>, <Thread(ThreadPoolExecutor-4_1, started daemon 139844764280576)>, <Thread(ThreadPoolExecutor-4_2, started daemon 139844755887872)>, <Thread(ThreadPoolExecutor-4_3, started daemon 139844596524800)>, <Thread(ThreadPoolExecutor-4_4, started daemon 139844588132096)>]
```

The three fail cases above share one common problem: lack of the log entries that reflects a correct execution of the topology state persistency. Let's take the last one to detail (all three have the same logic). Line 1 of fail 3 is the call right before persisting user request and returning to the user (https://github.com/kytos-ng/topology/blob/master/main.py#L273). Line 2 we can ignore because it belongs to previous request. Line 3 is called when the storehouse event is created to persist data. Line 4 is the return to the user request. Line 5 is the kytos shutdown process (hard shutdown, no further action/thread is executed).

As you can notice, there are missing events to actually persist data. The is because even though the persistency event was queued to be persisted, kytos shutdown occurred before the queue consumer thread gets the change to process that event.

The above scenario leads to a very undesired situation: Kytos returned success to the user, but the user request was not persisted (it was lost). The better approach would be one of: 1) kytos does not return success to the user before make sure the request was persisted; 2) if a success was returned to the user, kytos must guarantee the user request will not be lost.

Since 2 is almost impossible to accomplish due to unpredictable process crash situations, we better only return success to the user after making sure the request was persisted.

--- a/settings.py
+++ b/settings.py
@@ -5,8 +5,6 @@
 
 # Time (in seconds) to wait before setting a link as up
 LINK_UP_TIMER = 10
-# Time (in seconds) to wait retrieve box from storehouse
-BOX_RESTORE_TIMER = 0.1
 
 # Time (in seconds) to wait for a confirmation from storehouse
 # when retrieving or updating a box

--- a/settings.py
+++ b/settings.py
@@ -7,3 +7,10 @@
 LINK_UP_TIMER = 10
 # Time (in seconds) to wait retrieve box from storehouse
 BOX_RESTORE_TIMER = 0.1
+
+# Time (in seconds) to wait for a confirmation from storehouse
+# when retrieving or updating a box
+STOREHOUSE_TIMEOUT = 5.0
+
+# Time (in seconds) to sleep while waiting from storehouse (busy wait)
+STOREHOUSE_WAIT_INTERVAL = 0.05

--- a/storehouse.py
+++ b/storehouse.py
@@ -6,9 +6,6 @@ from kytos.core import log
 from kytos.core.events import KytosEvent
 from napps.kytos.topology import settings
 
-DEFAULT_BOX_RESTORE_TIMER = 0.1
-BOX_RESTORE_ATTEMPTS = 20
-
 
 class StoreHouse:
     """Class to handle storehouse."""
@@ -27,9 +24,6 @@ class StoreHouse:
         """Create a storehouse client instance."""
         self.controller = controller
         self.namespace = 'kytos.topology.status'
-        self.box_restore_timer = getattr(settings, 'BOX_RESTORE_TIMER',
-                                         DEFAULT_BOX_RESTORE_TIMER)
-
         if '_lock' not in self.__dict__:
             self._lock = threading.Lock()
         if 'box' not in self.__dict__:
@@ -40,9 +34,9 @@ class StoreHouse:
         """Return the persistence box data."""
         # Wait for box retrieve from storehouse
         i = 0
-        while not self.box and i < BOX_RESTORE_ATTEMPTS:
-            time.sleep(self.box_restore_timer)
-            i += 1
+        while not self.box and i < settings.STOREHOUSE_TIMEOUT:
+            time.sleep(settings.STOREHOUSE_WAIT_INTERVAL)
+            i += settings.STOREHOUSE_WAIT_INTERVAL
         if not self.box:
             error = 'Error retrieving persistence box from storehouse.'
             raise FileNotFoundError(error)

--- a/storehouse.py
+++ b/storehouse.py
@@ -1,4 +1,5 @@
 """Module to handle the storehouse."""
+import threading
 import time
 
 from kytos.core import log
@@ -29,6 +30,8 @@ class StoreHouse:
         self.box_restore_timer = getattr(settings, 'BOX_RESTORE_TIMER',
                                          DEFAULT_BOX_RESTORE_TIMER)
 
+        if '_lock' not in self.__dict__:
+            self._lock = threading.Lock()
         if 'box' not in self.__dict__:
             self.box = None
         self.list_stored_boxes()
@@ -96,20 +99,39 @@ class StoreHouse:
 
     def save_status(self, status):
         """Save the network administrative status using storehouse."""
-        self.box.data[status.get('id')] = status
+        storehouse_done = False
+        storehouse_error = None
 
-        content = {'namespace': self.namespace,
-                   'box_id': self.box.box_id,
-                   'data': self.box.data,
-                   'callback': self._save_status_callback}
+        def _storehouse_callback(event, data, error):
+            """Handle storehouse update result."""
+            # pylint: disable=unused-argument
+            nonlocal storehouse_done, storehouse_error
+            storehouse_done = True
+            storehouse_error = error
 
-        event = KytosEvent(name='kytos.storehouse.update', content=content)
-        self.controller.buffers.app.put(event)
+        with self._lock:
+            self.box.data[status.get('id')] = status
+            content = {'namespace': self.namespace,
+                       'box_id': self.box.box_id,
+                       'data': self.box.data,
+                       'callback': _storehouse_callback}
+            event = KytosEvent(name='kytos.storehouse.update',
+                               content=content)
+            self.controller.buffers.app.put(event)
 
-    def _save_status_callback(self, _event, data, error):
-        """Display the saved network status in the log."""
-        if error:
-            log.error(f'Can\'t update persistence box {data.box_id}.')
+            i = 0
+            while not storehouse_done and i < settings.STOREHOUSE_TIMEOUT:
+                time.sleep(settings.STOREHOUSE_WAIT_INTERVAL)
+                i += settings.STOREHOUSE_WAIT_INTERVAL
+
+            if not storehouse_done:
+                storehouse_error = 'Timeout while waiting for storehouse'
+
+            if storehouse_error:
+                log.error('Fail to update persistence box in '
+                          f'{self.namespace}.{self.box.box_id}. '
+                          f'Error: {storehouse_error}')
+                return
 
         log.info('Network administrative status saved in '
-                 f'{self.namespace}.{data.box_id}')
+                 f'{self.namespace}.{self.box.box_id}')

--- a/storehouse.py
+++ b/storehouse.py
@@ -28,6 +28,8 @@ class StoreHouse:
             self._lock = threading.Lock()
         if 'box' not in self.__dict__:
             self.box = None
+        self.storehouse_done = False
+        self.storehouse_error = None
         self.list_stored_boxes()
 
     def get_data(self):
@@ -93,39 +95,36 @@ class StoreHouse:
 
     def save_status(self, status):
         """Save the network administrative status using storehouse."""
-        storehouse_done = False
-        storehouse_error = None
-
-        def _storehouse_callback(event, data, error):
-            """Handle storehouse update result."""
-            # pylint: disable=unused-argument
-            nonlocal storehouse_done, storehouse_error
-            storehouse_done = True
-            storehouse_error = error
-
         with self._lock:
+            self.storehouse_done = False
+            self.storehouse_error = None
             self.box.data[status.get('id')] = status
             content = {'namespace': self.namespace,
                        'box_id': self.box.box_id,
                        'data': self.box.data,
-                       'callback': _storehouse_callback}
+                       'callback': self._save_status_callback}
             event = KytosEvent(name='kytos.storehouse.update',
                                content=content)
             self.controller.buffers.app.put(event)
 
             i = 0
-            while not storehouse_done and i < settings.STOREHOUSE_TIMEOUT:
+            while not self.storehouse_done and i < settings.STOREHOUSE_TIMEOUT:
                 time.sleep(settings.STOREHOUSE_WAIT_INTERVAL)
                 i += settings.STOREHOUSE_WAIT_INTERVAL
 
-            if not storehouse_done:
-                storehouse_error = 'Timeout while waiting for storehouse'
+            if not self.storehouse_done:
+                self.storehouse_error = 'Timeout while waiting for storehouse'
 
-            if storehouse_error:
+            if self.storehouse_error:
                 log.error('Fail to update persistence box in '
                           f'{self.namespace}.{self.box.box_id}. '
-                          f'Error: {storehouse_error}')
+                          f'Error: {self.storehouse_error}')
                 return
 
         log.info('Network administrative status saved in '
                  f'{self.namespace}.{self.box.box_id}')
+
+    def _save_status_callback(self, _event, _data, error):
+        """Handle storehouse update result."""
+        self.storehouse_done = True
+        self.storehouse_error = error

--- a/tests/unit/test_storehouse.py
+++ b/tests/unit/test_storehouse.py
@@ -24,8 +24,10 @@ class TestStoreHouse(TestCase):
 
         self.napp = StoreHouse(get_controller_mock())
 
+    @patch('time.sleep', return_value=None)
+    @patch('napps.kytos.topology.storehouse.settings')
     @patch('napps.kytos.topology.storehouse.StoreHouse.get_stored_box')
-    def test_get_data(self, mock_get_stored_box):
+    def test_get_data(self, mock_get_stored_box, mock_settings, mock_sleep):
         """Test get_data."""
         mock_box = MagicMock()
         box_data = MagicMock()
@@ -39,6 +41,14 @@ class TestStoreHouse(TestCase):
         response = self.napp.get_data()
         self.assertEqual(response.data, "box")
 
+        # test timeout
+        mock_settings.STOREHOUSE_TIMEOUT = 2
+        mock_settings.STOREHOUSE_WAIT_INTERVAL = 0.5
+        self.napp.box = None
+        with self.assertRaises(FileNotFoundError):
+            self.napp.get_data()
+        self.assertEqual(mock_sleep.call_count, 4)
+
     @patch('napps.kytos.topology.storehouse.KytosEvent')
     @patch('kytos.core.buffers.KytosEventBuffer.put')
     def test_create_box(self, *args):
@@ -47,6 +57,16 @@ class TestStoreHouse(TestCase):
         self.napp.create_box()
         mock_event.assert_called()
         mock_buffers_put.assert_called()
+
+    # pylint: disable = protected-access
+    @patch('napps.kytos.topology.storehouse.log')
+    def test_create_box_callback(self, mock_log):
+        """Test _create_box_callback."""
+        mock_data = MagicMock()
+        self.napp._create_box_callback('event', mock_data, None)
+        mock_log.error.assert_not_called()
+        self.napp._create_box_callback('event', mock_data, 'error')
+        mock_log.error.assert_called()
 
     # pylint: disable = protected-access
     @patch('napps.kytos.topology.storehouse.StoreHouse.get_stored_box')
@@ -76,12 +96,47 @@ class TestStoreHouse(TestCase):
         mock_event.assert_called()
         mock_buffers_put.assert_called()
 
-    @patch('napps.kytos.topology.storehouse.KytosEvent')
+    # pylint: disable = protected-access
+    @patch('napps.kytos.topology.storehouse.log')
+    def test_get_box_callback(self, mock_log):
+        """Test _get_box_callback."""
+        mock_data = MagicMock()
+        self.napp._get_box_callback('event', mock_data, None)
+        mock_log.error.assert_not_called()
+        self.napp._get_box_callback('event', mock_data, 'error')
+        mock_log.error.assert_called()
+
+    @patch('time.sleep', return_value=None)
+    @patch('napps.kytos.topology.storehouse.log')
     @patch('kytos.core.buffers.KytosEventBuffer.put')
     def test_save_status(self, *args):
         """Test save_status."""
-        (mock_buffers_put, mock_event) = args
+        (mock_buffers_put, mock_log, mock_sleep) = args
         mock_status = MagicMock()
+        self.napp.box = MagicMock()
+
+        # case 1: successfull return
+        def buffers_put_side_effect_1(event):
+            event.content['callback']('event', 'data', None)
+        mock_buffers_put.side_effect = buffers_put_side_effect_1
         self.napp.save_status(mock_status)
-        mock_event.assert_called()
         mock_buffers_put.assert_called()
+        mock_sleep.assert_not_called()
+        mock_log.info.assert_called()
+
+        # case 2: error from storehouse
+        def buffers_put_side_effect_2(event):
+            event.content['callback']('event', 'data', 'error')
+        mock_buffers_put.side_effect = buffers_put_side_effect_2
+        self.napp.save_status(mock_status)
+        mock_buffers_put.assert_called()
+        mock_sleep.assert_not_called()
+        mock_log.error.assert_called()
+
+        # case 3: timeout
+        mock_log.error.call_count = 0
+        mock_buffers_put.side_effect = [None]
+        self.napp.save_status(mock_status)
+        mock_buffers_put.assert_called()
+        mock_sleep.assert_called()
+        mock_log.error.assert_called()

